common/src/main/java/tachyon/client/FileSystemMasterClient.java

```diff
@@ -0,0 +1,397 @@
+/*
+ * Licensed to the University of California, Berkeley under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.client;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import tachyon.Constants;
+import tachyon.MasterClientBase;
+import tachyon.conf.TachyonConf;
+import tachyon.thrift.BlockInfoException;
+import tachyon.thrift.DependencyDoesNotExistException;
+import tachyon.thrift.DependencyInfo;
+import tachyon.thrift.FileAlreadyExistException;
+import tachyon.thrift.FileBlockInfo;
+import tachyon.thrift.FileDoesNotExistException;
+import tachyon.thrift.FileInfo;
+import tachyon.thrift.FileSystemMasterService;
+import tachyon.thrift.InvalidPathException;
+import tachyon.thrift.SuspectedFileSizeException;
+
+/**
+ * The FileSystemMaster client, for clients.
+ *
+ * Since thrift clients are not thread safe, this class is a wrapper to provide thread safety.
```

Comment:
Not 100% accurate, because this wrapper class is more than providing thread safety, but also implement retry logic (currently an infinite loop) to handle thrift exceptions.